File: Coding Ninja/3sum.cpp
#include <bits/stdc++.h> 
vector<vector<int>> findTriplets(vector<int>nums, int n, int K) {    //Defining function findTriplets
	// Write your code here.
	if(n<3)return {};    // check size of array

	vector<vector<int>> triplets;    // initialize an array of triplets

    sort(nums.begin(), nums.end());    // sort the array

	for(int i = 0; i<n-2; i++){    //iterate over array
        int start = i+1;    // initialize start variable
        int end = n-1;    // initialize end variable
        while(start<end){    // check if start is less than end
                
            int sum = nums[i] + nums[start] + nums[end];    //calculate sum
            if(sum == K){    // check if sum is equal to K
                triplets.push_back({nums[i], nums[start], nums[end]});    // push triplet in triplets array
                start++;    // increment start
                end--;    // decrement end
                while(start<n && nums[start] == nums[start-1]){    // check if start is less than n and if element at start and start-1 is equal
					start++;    //increment start
				}
					
                while(end>i && nums[end] == nums[end+1]){    // check if end is greater than i and if element at end and end+1 is equal
					end--;    //decrement end
				}	
            }
			else if(sum>K){    // check if sum is greater than K
                end--;    // decrement end
            } 
			else {    // else body
              start++;    // increment start
            }
        }
        while (i < n - 1 && nums[i] == nums[i + 1])    // check if i is less than n-1 and if element at i and i+1 is equal
          i++;    // increment i
    }
    return triplets;    // return triplets array
}

File: Coding Ninja/3sum.py
from os import *    # import os module

from sys import *    # import sys module

from collections import *    # import collections module

from math import *    # import math module

from sys import stdin, stdout, setrecursionlimit    # import stdin, stdout and setrecursionlimit

def findTriplets(arr, n, k):    # findTriplets function to find triplets, which sums to k

    arr.sort()    # sort the array

    res=[]    # initialize result array

    for i in range(n-2):    # iterate over n-2 elements of array

        if arr[i]==arr[i-1] and i>0:    # check if value at i is same as value at i-1

            continue    # continue the iteration

        left=i+1    # initialize left pointer to i+1

        right=n-1    # initialize right pointer to n-1

        while left<right:    # iterate till left is less than right

            curr_sum=arr[i]+arr[left]+arr[right]    # curr_sum is the sum of values at i, left and right

            if curr_sum==k and [arr[i],arr[left],arr[right]] not in res:    # check if curr_sum is equal to k and triplet is not there in res array

                res.append([arr[i],arr[left],arr[right]])    # append the triplet to res array

                left+=1    # increment left by 1

                right-=1    # decrement right by 1

                

            elif curr_sum>k:    # check if curr_sum is greater than k

                right-=1    # decrement right by 1

            else:    # else body

                left+=1    # increment left by 1

    return res    # return res array

# Taking input using fast I/0.    # Taking input using fast I/0
def takeInput():    # takeInput function
    N = int(stdin.readline())    # read N from input
    arr = list(map(int, stdin.readline().strip().split(" ")))    # read array
    K = int(stdin.readline())    # read K from input
    return N, arr,K    # return N, arr and K

tc = int(input())    # read test case count
while tc > 0:    # iterate till tc is greater than 0
    N, arr,K = takeInput()    # takeInput
    ans = findTriplets(arr, N,K)    # find triplets which sum to K

    if len(ans) == 0:    # check if ans is empty
        stdout.write("-1\n")    # write -1 on console

    else:    # else body
        for i in ans:    # iterate over ans
            i.sort()    # sort the triplet
            stdout.write(str(i[0]) + " " + str(i[1]) + " " + str(i[2]) + "\n")    # write the triplet on console

    tc -= 1    # decrement tc by 1
